Adding Profile Store functionality

murilommen opened the pull request:
## Description

This repository needs a Profile store wrapper, that will manage `get` and `write` for structured timestamp-based queries

- [x] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

murilommen committed:
adding get method

murilommen committed:
store write logic

murilommen committed:
get and write refactored

murilommen committed:
Merge remote-tracking branch 'origin/mainline' into dev/murilommen/profile-store

murilommen committed:
unit tests in

details: added scope-based unit tests for local_store
and utils of the API store in general

Signed-off-by: murilommen <murilommen@gmail.com>

murilommen committed:
Merge remote-tracking branch 'origin/mainline' into dev/murilommen/profile-store

murilommen committed:
adding docstrings

naddeoa commented on python/examples/advanced/Profile_Store/LocalStore/constraints_example.py:
```diff
@@ -0,0 +1,23 @@
+# 1. Configure Store
+store = ProfileStore(base_name="test_constraints")
+now_config = DateConfig(start_date=datetime.now())
+reference_config = DateConfig(start_date=datetime.now() - timedelta(days=7), end_date=datetime.now())
```
Rather than a config, it might make more sense to call this something like a `StoreQuery`. It isn't really configuring anything, its just a way to ask for profiles right? Then you would have `DateQuery`, `ProfileIdQuery`, etc.

murilommen replied:
to answer myself, I think I got somewhere:
```python
query_date = DateQuery(start, end)
query_id = ProfileIdQuery(id)
# ---
store.get(any_query: StoreQuery)
```
internally, `store.get` will always call `StoreQuery.query()`, which is the abstract method. sounds good?

naddeoa commented on python/examples/advanced/Profile_Store/LocalStore/constraints_example.py:
```diff
@@ -0,0 +1,23 @@
+now_config = DateConfig(start_date=datetime.now())
+reference_config = DateConfig(start_date=datetime.now() - timedelta(days=7), end_date=datetime.now())
+
+# 2. Get the profile
+profile_view = store.get(date_config=now_config)
+reference_profile = store.get(date_config=reference_config)
```
Is it appropriate to use polymorphism here in python? The `date_config` would just be `config` (or `query` if my other comment) and people would just pass in whatever type of `StoreQuery` they want to use.

murilommen replied:
yes! makes total sense

naddeoa commented on python/whylogs/api/logger/rolling.py:
```diff
@@ -126,6 +129,10 @@ def __init__(
+    def append_store(self, store: ProfileStore) -> None:
+        if store is not None:
+            self.store = store
+
```
This looks more like a `set_store` than an `append_store`. That would imply we maintain a list of stores. I think its actually an ok idea to have a list of stores but it doesn't look like that's in yet. What was the intent?

murilommen replied:
yeah, the intent was to enable the same behavior of the writers, to write for every store on the store_list. will make sure to fix that!

naddeoa commented on python/whylogs/api/store/profile_store.py:
```diff
@@ -1,51 +1,26 @@
     @abstractmethod
-    def query(self, *args: Any, **kwargs: Any) -> List[DatasetProfile]:
+    def write(self, profile: Optional[DatasetProfileView]) -> None:
         pass
```
I guess you were thinking about `query` at one point. Maybe it would make more sense to have `get` be key based and then allow for a separate `query` api that lets you query based on date ranges or other properties? Might even make sense to make it 

```python
def query(self, query: List[StoreQuery]) -> List[DatasetProfile]
```

Then the api would allow for people to specify multiple criteria, which would allow them to do `or` with multiple date ranges.

murilommen replied:
actually the `query` method was from Jamie's stubs 😄 
so if I have 2 profiles written today, using `base_name-{timestamp}-{random_string}`, how can I refer to it in order to get it? since we discussed that `id` would only be the `base_name` at the moment

jamie256 replied:
we can add query later, starting with this tighter scope looks reasonable.

naddeoa commented on python/whylogs/api/store/local.py:
```diff
@@ -0,0 +1,103 @@
+        self._writer = LocalWriter(base_dir=self._default_path)
+
+        if not os.path.isdir(os.path.join(self._default_path, self.base_name)):
+            os.makedirs(os.path.join(self._default_path, self.base_name))
```
might just move this to the write method so it always happens before writes and it doesn't happen if nothing is ever written. I can imagine doing some local tests on a loop or something, manually deleting the profile dir to test something, then having to restart the app because it can't remake the dir outside of startup.

naddeoa commented on python/whylogs/api/store/local.py:
```diff
@@ -0,0 +1,103 @@
+            files_list.extend(glob(f"{base_directory}/profile_{date}*.bin"))
+        if len(files_list) > 0:
+            logger.debug(f"Profiles found! Number of profiles is {len(files_list)}")
+            profiles_list = [why.read(file).view() for file in files_list]
```
I guess this comment applies more for the API that just returns views, Noticed that line below does the merge. A similar comment applies to the merge though: we can read them one at a time and merge them right away to avoid memory issues, which would just mean updating `x.merge` to `read && x.merge`

murilommen replied:
as easy as changing it to parenthesis. sometimes i really love Python :D

naddeoa commented on python/whylogs/api/store/local.py:
```diff
@@ -0,0 +1,103 @@
+    def _get_profile_timestamp() -> str:
+        now = datetime.utcnow()
+        return f"profile_{now.date()}_{now.hour}:{now.minute}:{now.second}_{random_string()}.bin"
+
```
More like `_get_profile_filename` right? Might be a python question here, why bother making private methods static rather than just making this a function in the module above?

murilommen replied:
renamed it! 
about the private vs. function, i rather keep this method inside the class because I only want local store to use this, not other stuff. and since it doesn't rely on any of the object's attrs, i made it static.

naddeoa commented on python/whylogs/api/store/local.py:
```diff
@@ -0,0 +1,103 @@
+        return profile
+
+    def write(self, profile: Optional[DatasetProfileView]) -> None:
+        profile_dest = os.path.join(self._default_path, self.base_name, self._get_profile_timestamp())
```
Why `Optional`?

murilommen commented on python/examples/advanced/Profile_Store/LocalStore/generate_data.py:
```diff
@@ -0,0 +1,21 @@
+logger = why.logger(mode="rolling", interval=10, when="S")
+logger.append_store(store=ProfileStore(base_name="test_constraints"))
+
+
```
and users could always use the regular LocalStore api, as in:
```python
store = LocalStore(base_name="test_constraints")
store.write(profile=some_profile)

store.get(config=DateConfig(some_datas_timestamp))
```

murilommen replied:
i'm here assuming `logger.log()` and `store.get()` will happen on different processes. 1) capture 2) compare

murilommen committed:
modified DateConfig behaviour

details: before, DateConfig was acting just as a placeholder for dates
Now, introducing the concept of StoreQueries which will
standardize the way we fetch DatasetProfileViews from the ProfileStore

murilommen committed:
Merge branch 'mainline' into dev/murilommen/profile-store

murilommen commented on python/whylogs/api/store/local.py:
```diff
@@ -0,0 +1,103 @@
+
+    In order to instantiate the object, you will need to define a base_name,
+    which is related to the name of the model or dataset you're profiling.
+    To properly use the LocalStore to generate files, you should append it
```
adding in a footer note on `base_name`. check if it explains better what it's meant to be

murilommen committed:
parameters renaming, docs and yield vs. return

 details:
 * DateConfig was renamed to DateQuery, but some functionality was brought back
 to the LocalStore after the latest commit.
 DateQuery is mainly something to allocate start_date and end_date properly
 * More docs on base_name
 * Profiles list now is a generator in order to make it less memory-consuming

naddeoa commented on python/whylogs/api/store/local_store.py:
```diff
@@ -0,0 +1,111 @@
+# TODO how to prevent users from creating a huge list?
+# should the profile store be aware of its min/max dates?
+
+
```
I think we can get around this by avoiding loading everything into memory at once if the concern is just not letting the host go oom. To do that, we'd have to have make it possible for the store to use that yield iterator in the reduce.

murilommen replied:
the only catch is that we are still building the `n_days` list in memory

naddeoa commented on python/whylogs/api/store/profile_store.py:
```diff
@@ -1,51 +1,28 @@
-        schema: Optional[DatasetSchema] = None,
-    ) -> None:
+    def get(self, query: DateQuery) -> DatasetProfileView:
         pass
```
Just occurred to me, but I guess we don't have to worry too much about having a common base query type, we can just use `Union[DateQuery, ProfileIdQuery, etc]`

murilommen replied:
sounds good! i will resolve this for now and when we have a more mature idea on a `ProfileIdQuery` class we will add an `Union` type there ;)

naddeoa commented on python/whylogs/api/store/profile_store.py:
```diff
@@ -1,51 +1,28 @@
+    # TODO return datetimes instead of strings
+    def _get_dates(self, query: DateQuery) -> List[str]:
+        dates = []
+        for single_date in self._get_date_range(query=query):
```
Not sure if we're waiting to do this or not?

murilommen replied:
yes, we should! my mistake to leave this behind, but luckily you caught it

murilommen commented on python/whylogs/api/store/query.py:
```diff
@@ -0,0 +1,18 @@
+class DateQuery:
+    start_date: datetime
+    end_date: Optional[datetime] = field(default=None)
+
```
it goes on the post_init method. if no end_date, `end_date == start_date`
also note there that if `end_date < start_date`, then i switch them

murilommen committed:
store._get_dates returns datetime

details: get_dates returning date objects makes it clearer to what the
method is doing and also allows for the LocalStore to handle it
specifically to how it needs.
also removed TODO for granular queries to limit the scope of this PR for now

murilommen committed:
store now has list + profile_name query

details: users are able to list existing profiles with store.list()
it is now also possible to query entire profile directories
using the ProfileNameQuery.

murilommen committed:
renaming profile -> profile_view

details: profile argument now is profile view for clarity
also fixed some failing tests

murilommen committed:
fix typing errors

murilommen committed:
close rolling logger

murilommen committed:
adding localstore notebook + fixing append method

murilommen committed:
raising the notebook tests timeout

murilommen commented on python/examples/advanced/Profile_Store/LocalStore/LocalStore_with_Constraints.ipynb:
```diff
@@ -0,0 +1,1042 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "And since we have an empty list returned, it means that we haven't used the profile store in this location so far. But we can already check that a new directory called `profile_store` was created on our working directory:"
```
good catch! will re-run things to clear it up :)

murilommen committed:
fix notebook output

jamie256 commented on python/whylogs/api/store/utils/random_strings.py:
```diff
@@ -0,0 +1,7 @@
+
+
+def random_string(length: int = 7):
+    letters = string.ascii_lowercase
```
do the filenames need to be readable and short?

I'm concerned this could generate inappropriate/offensive filenames, and if this is not for human readability, then I'd prefer something like a uuid string here instead.

murilommen replied:
done!

jamie256 reviewed (commented):
Can we change the random file name generation a bit?

murilommen committed:
changing random_strings.py generation to uuid4

details: from preventing ourselves to potentially write
weird/offensive file names in the future with random,
going now with uuid4, as suggested by @jamie256

jamie256 reviewed (approved):
LGTM!
